core/src/main/scala/kafka/server/instrumentation/ProduceRequestInstrumentation.scala

```diff
@@ -0,0 +1,193 @@
+package kafka.server.instrumentation
+
+import com.typesafe.scalalogging.Logger
+import kafka.cluster.PendingShrinkIsr
+import kafka.network.RequestChannel
+import kafka.server.instrumentation.ProduceRequestInstrumentation.Stage
+import kafka.server.instrumentation.ProduceRequestInstrumentationLogger.Config
+import kafka.server.{BrokerReconfigurable, KafkaConfig, ReplicaManager}
+import kafka.utils.CoreUtils.nanosToMs
+import kafka.utils.Logging
+import org.apache.kafka.common.TopicPartition
+import org.apache.kafka.common.config.ConfigException
+import org.apache.kafka.common.utils.Time
+
+import scala.collection.mutable
+
+object ProduceRequestInstrumentation {
+  /**
+   * Symbols to be used for marking at each instrumentation stage
+   */
+  object Stage extends Enumeration {
+    // Define stages here
+    val Init,
+        Authorization,
+        BeginAppendRecords,
+        AppendToLocalLog,
+        ProcessAppendToLocalLogStatus,
+        PrepareDelayedProduce,
+        EnqueueDelayedProduce,
+        BeginResponseCallback,
+        ResponseThrottling,
+        Finish = Value
```

Comment:
What this is missing (since the details happen on other machines) is the replica-fetcher delays. We want to flag high fetch delays and break them down similarly, then (ideally) correlate them with long delays on the producer side. (I can't comment on nonexistent code, so it's going here.)

That's an additional problem with the randomness thing, btw: if both sides have 50% logRatios, only 12.5% of the logged requests will have full info in the RF=3 case, and I think krc is RF=4, which would be even worse (6.25%). (Of course, if it's a replica-fetching delay, probably just one replica would be affected, so realistically you get "full" info 25% of the time. But you'll never know for sure since you simply don't have the rest of the info.)

Reply:
This is a even more complicated problem I think.

In Kafka's design, fetch is totally decoupled from the produce request.
For example:
We have 3 produce requests to a partition, sending 5 messages each, waiting for a follower to complete fetching.
And a follower, after some GC pause, fetches the 15 messages at once and unblocks the 3 produce requests altogether.  In that case, we will have to sort of "attribute" the request to the message body to have such trace.

That would require a more complicated design & discussion.

Reply:
I feel like it should be similar to what you've done on the producer half, just that we'll want to include offset info on both sides. Then if, say, a single replica spends a lot of time on the fetch (for whatever reason), it can log its accumulated breakdown of timings and also the offset after the batch is written (or better, the offset range of the full batch), and with that we'll be able to associate it with similarly marked slow produce requests even if the latter's offset range is a subset of the former's. Other than the manual aspect of linking such logs across different hosts with non-identical offset ranges, do you see issues with that approach?

Reply:
We can add that to future work.

A concern for me that is the amount of instrumentation & logging; say RF=4 as in KRC, the instrumentation here amount is on the leader, and in theory that's going to be 3X instrumentation happening.  We may need to tweak lower on the sample rate, but that may worsen if we're trying to debug a particular request under the sample-and-match case.

Reply:
On the follower half, to be accurate, what I can think of is a similar approach, but instead of thresholding on the request completion time, thresholding on the timestamp diff of (earliest message in fetched batch event time ~ complete time)